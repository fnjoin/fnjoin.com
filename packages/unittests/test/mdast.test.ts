import { expect, test } from "@jest/globals";
import { toHtml } from "hast-util-to-html";
import {
    directiveFromMarkdown,
    directiveToMarkdown,
} from "mdast-util-directive";
import { fromMarkdown } from "mdast-util-from-markdown";
import { toHast } from "mdast-util-to-hast";
import { toMarkdown } from "mdast-util-to-markdown";
import { directive } from "micromark-extension-directive";

/**
 * @typedef {import('mdast-util-to-hast')}
 */

test("mdast", () => {
    // read a file synchronously
    // const doc = fs.readFileSync("example.md", "utf8");
    // const doc = await fs.readF("example.md");
    const doc = `A lovely language know as :abbr[HTML]{title="HyperText Markup Language"}.`;

    const tree = fromMarkdown(doc, {
        extensions: [directive()],
        mdastExtensions: [directiveFromMarkdown()],
    });

    expect(removePositionKeys(tree)).toMatchSnapshot();
    // console.log(tree);
    // console.log(JSON.stringify(removePositionKeys(tree), null, 2));

    toMarkdown(tree, { extensions: [directiveToMarkdown()] });

    // console.log(out);
});

test("mdast has text directives but how flexible are they?", () => {
    const doc = `A lovely language known as :anything[stuff]{title="This is something"}.`;

    const tree = fromMarkdown(doc, {
        extensions: [directive()],
        mdastExtensions: [directiveFromMarkdown()],
    });

    expect(removePositionKeys(tree)).toMatchSnapshot();

    const out = toMarkdown(tree, { extensions: [directiveToMarkdown()] });

    expect(out).toMatchSnapshot();
});

interface AnyObject {
    [key: string]: any;
}
function removePositionKeys(obj: AnyObject): AnyObject {
    if (Array.isArray(obj)) {
        return obj.map(removePositionKeys);
    } else if (typeof obj === "object" && obj !== null) {
        const { position, ...rest } = obj; // Destructure to remove `position` key if it exists
        return Object.keys(rest).reduce((acc, key) => {
            // Recursively apply to all properties
            acc[key] = removePositionKeys(rest[key]);
            return acc;
        }, {} as AnyObject);
    }
    return obj;
}

test("does mdast have block level elements like pandoc?", () => {
    const doc = `
:::spoiler{key="value"}
He dies.
:::
    `;

    const tree = fromMarkdown(doc, {
        extensions: [directive()],
        mdastExtensions: [directiveFromMarkdown()],
    });

    // take a snapshot of tree and expect
    expect(removePositionKeys(tree)).toMatchSnapshot();

    // const out = toMarkdown(tree, { extensions: [directiveToMarkdown()] });
});

test("mdast to html", () => {
    const doc = `
Maybe needs to be longer than a single thing?

:::spoiler{key="value" class="foo"}
He dies.
:::

I don't know.

Some \`backticks\` for inline code.
    `;

    const tree = fromMarkdown(doc, {
        extensions: [directive()],
        mdastExtensions: [directiveFromMarkdown()],
    });

    const h = toHast(tree, {
        handlers: {
            containerDirective(state, node) {
                return {
                    type: "element",
                    tagName: "div",
                    properties: {
                        id: node.name,
                        anything: "foo",
                        className: ["blah", node.attributes.key],
                    },
                    children: state.all(node),
                };
            },
        },
    });
    const html = toHtml(h);

    expect(tree).toMatchSnapshot();
    expect(h).toMatchSnapshot();
    expect(html).toMatchSnapshot();
});
