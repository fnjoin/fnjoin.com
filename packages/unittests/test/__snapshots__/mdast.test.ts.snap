// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does mdast have block level elements like pandoc? 1`] = `
{
  "children": [
    {
      "attributes": {
        "key": "value",
      },
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "He dies.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "name": "spoiler",
      "type": "containerDirective",
    },
  ],
  "type": "root",
}
`;

exports[`mdast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "A lovely language know as ",
        },
        {
          "attributes": {
            "title": "HyperText Markup Language",
          },
          "children": [
            {
              "type": "text",
              "value": "HTML",
            },
          ],
          "name": "abbr",
          "type": "textDirective",
        },
        {
          "type": "text",
          "value": ".",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`mdast has text directives but how flexible are they? 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "A lovely language known as ",
        },
        {
          "attributes": {
            "title": "This is something",
          },
          "children": [
            {
              "type": "text",
              "value": "stuff",
            },
          ],
          "name": "anything",
          "type": "textDirective",
        },
        {
          "type": "text",
          "value": ".",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`mdast has text directives but how flexible are they? 2`] = `
"A lovely language known as :anything[stuff]{title="This is something"}.
"
`;

exports[`mdast to html 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 46,
              "line": 2,
              "offset": 46,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "Maybe needs to be longer than a single thing?",
        },
      ],
      "position": {
        "end": {
          "column": 46,
          "line": 2,
          "offset": 46,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
    {
      "attributes": {
        "class": "foo",
        "key": "value",
      },
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 9,
                  "line": 5,
                  "offset": 92,
                },
                "start": {
                  "column": 1,
                  "line": 5,
                  "offset": 84,
                },
              },
              "type": "text",
              "value": "He dies.",
            },
          ],
          "position": {
            "end": {
              "column": 9,
              "line": 5,
              "offset": 92,
            },
            "start": {
              "column": 1,
              "line": 5,
              "offset": 84,
            },
          },
          "type": "paragraph",
        },
      ],
      "name": "spoiler",
      "position": {
        "end": {
          "column": 4,
          "line": 6,
          "offset": 96,
        },
        "start": {
          "column": 1,
          "line": 4,
          "offset": 48,
        },
      },
      "type": "containerDirective",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 14,
              "line": 8,
              "offset": 111,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 98,
            },
          },
          "type": "text",
          "value": "I don't know.",
        },
      ],
      "position": {
        "end": {
          "column": 14,
          "line": 8,
          "offset": 111,
        },
        "start": {
          "column": 1,
          "line": 8,
          "offset": 98,
        },
      },
      "type": "paragraph",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 10,
              "offset": 118,
            },
            "start": {
              "column": 1,
              "line": 10,
              "offset": 113,
            },
          },
          "type": "text",
          "value": "Some ",
        },
        {
          "position": {
            "end": {
              "column": 17,
              "line": 10,
              "offset": 129,
            },
            "start": {
              "column": 6,
              "line": 10,
              "offset": 118,
            },
          },
          "type": "inlineCode",
          "value": "backticks",
        },
        {
          "position": {
            "end": {
              "column": 34,
              "line": 10,
              "offset": 146,
            },
            "start": {
              "column": 17,
              "line": 10,
              "offset": 129,
            },
          },
          "type": "text",
          "value": " for inline code.",
        },
      ],
      "position": {
        "end": {
          "column": 34,
          "line": 10,
          "offset": 146,
        },
        "start": {
          "column": 1,
          "line": 10,
          "offset": 113,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 11,
      "offset": 151,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdast to html 2`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 46,
              "line": 2,
              "offset": 46,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "Maybe needs to be longer than a single thing?",
        },
      ],
      "position": {
        "end": {
          "column": 46,
          "line": 2,
          "offset": 46,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
    {
      "type": "text",
      "value": "
",
    },
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 9,
                  "line": 5,
                  "offset": 92,
                },
                "start": {
                  "column": 1,
                  "line": 5,
                  "offset": 84,
                },
              },
              "type": "text",
              "value": "He dies.",
            },
          ],
          "position": {
            "end": {
              "column": 9,
              "line": 5,
              "offset": 92,
            },
            "start": {
              "column": 1,
              "line": 5,
              "offset": 84,
            },
          },
          "properties": {},
          "tagName": "p",
          "type": "element",
        },
      ],
      "properties": {
        "anything": "foo",
        "className": [
          "blah",
          "value",
        ],
        "id": "spoiler",
      },
      "tagName": "div",
      "type": "element",
    },
    {
      "type": "text",
      "value": "
",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 14,
              "line": 8,
              "offset": 111,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 98,
            },
          },
          "type": "text",
          "value": "I don't know.",
        },
      ],
      "position": {
        "end": {
          "column": 14,
          "line": 8,
          "offset": 111,
        },
        "start": {
          "column": 1,
          "line": 8,
          "offset": 98,
        },
      },
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
    {
      "type": "text",
      "value": "
",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 10,
              "offset": 118,
            },
            "start": {
              "column": 1,
              "line": 10,
              "offset": 113,
            },
          },
          "type": "text",
          "value": "Some ",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 17,
                  "line": 10,
                  "offset": 129,
                },
                "start": {
                  "column": 6,
                  "line": 10,
                  "offset": 118,
                },
              },
              "type": "text",
              "value": "backticks",
            },
          ],
          "position": {
            "end": {
              "column": 17,
              "line": 10,
              "offset": 129,
            },
            "start": {
              "column": 6,
              "line": 10,
              "offset": 118,
            },
          },
          "properties": {},
          "tagName": "code",
          "type": "element",
        },
        {
          "position": {
            "end": {
              "column": 34,
              "line": 10,
              "offset": 146,
            },
            "start": {
              "column": 17,
              "line": 10,
              "offset": 129,
            },
          },
          "type": "text",
          "value": " for inline code.",
        },
      ],
      "position": {
        "end": {
          "column": 34,
          "line": 10,
          "offset": 146,
        },
        "start": {
          "column": 1,
          "line": 10,
          "offset": 113,
        },
      },
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 11,
      "offset": 151,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdast to html 3`] = `
"<p>Maybe needs to be longer than a single thing?</p>
<div id="spoiler" anything="foo" class="blah value"><p>He dies.</p></div>
<p>I don't know.</p>
<p>Some <code>backticks</code> for inline code.</p>"
`;
